circuitx/navigation/src/commonMain/kotlin/com/slack/circuitx/navigation/intercepting/CircuitInterceptingNavigator.kt

```diff
@@ -0,0 +1,199 @@
+) : Navigator by delegate {
+
+  override fun goTo(screen: Screen): Boolean {
+    for (interceptor in interceptors) {
+      when (val interceptorResult = interceptor.goTo(screen)) {
+        is InterceptorResult.Skipped -> continue
+        is InterceptorResult.Success -> {
+          if (interceptorResult.consumed) return true
+        }
+        is InterceptorResult.Failure -> {
+          notifier?.goToInterceptorFailure(interceptorResult)
+          if (interceptorResult.consumed) return false
+        }
+        is InterceptorGoToResult.Rewrite -> {
+          // Recurse in case another interceptor wants to intercept the new screen.
+          return goTo(interceptorResult.screen)
+        }
+      }
+    }
+    eventListeners.forEach { it.goTo(screen) }
+    return delegate.goTo(screen)
+  }
+
+  override fun pop(result: PopResult?): Screen? {
+    val backStack = peekBackStack()
+    for (interceptor in interceptors) {
+      when (val interceptorResult = interceptor.pop(backStack, result)) {
+        is InterceptorResult.Skipped -> continue
+        is InterceptorResult.Success -> {
+          if (interceptorResult.consumed) return null
+        }
+        is InterceptorResult.Failure -> {
+          notifier?.popInterceptorFailure(interceptorResult)
+          if (interceptorResult.consumed) return null
+        }
+      }
+    }
+    eventListeners.forEach { it.pop(backStack, result) }
+    return delegate.pop(result)
+  }
+
+  override fun resetRoot(
+    newRoot: Screen,
+    saveState: Boolean,
+    restoreState: Boolean,
+  ): ImmutableList<Screen> {
+    for (interceptor in interceptors) {
+      when (val interceptorResult = interceptor.resetRoot(newRoot, saveState, restoreState)) {
+        is InterceptorResult.Skipped -> continue
+        is InterceptorResult.Success -> {
+          if (interceptorResult.consumed) return persistentListOf()
+        }
+        is InterceptorResult.Failure -> {
+          notifier?.rootResetInterceptorFailure(interceptorResult)
+          if (interceptorResult.consumed) return persistentListOf()
+        }
+        is InterceptorResetRootResult.Rewrite -> {
+          // Recurse in case another interceptor wants to intercept the new screen.
+          return resetRoot(
+            interceptorResult.screen,
+            interceptorResult.saveState,
+            interceptorResult.restoreState,
+          )
+        }
+      }
+    }
+    eventListeners.forEach { it.resetRoot(newRoot, saveState, restoreState) }
+    return delegate.resetRoot(newRoot, saveState, restoreState)
+  }
+
+  /** Notifies of [CircuitNavigationInterceptor] failures. Useful for logging or analytics. */
+  @Immutable
+  public interface FailureNotifier {
+
+    /**
+     * Notifies of a [InterceptorResult.Failure] from a [CircuitNavigationInterceptor] during a
+     * [CircuitNavigationInterceptor.goTo].
+     */
+    public fun goToInterceptorFailure(interceptorResult: InterceptorResult.Failure)
```

Comment:
nit: `goToFailure`? Since this is already being used in the context of an interceptor